aptos-move/framework/aptos-framework/sources/zkid.move

```diff
@@ -0,0 +1,123 @@
+        initialize(fx, vk, constants)
+    }
+
+    public fun new_groth16_verification_key(alpha_g1: vector<u8>, beta_g2: vector<u8>, gamma_g2: vector<u8>, delta_g2: vector<u8>, gamma_abc_g1: vector<vector<u8>>): Groth16VerificationKey {
+        Groth16VerificationKey {
+            alpha_g1,
+            beta_g2,
+            gamma_g2,
+            delta_g2,
+            gamma_abc_g1,
+        }
+    }
+
+    /// Returns the Groth16 VK for our devnet deployment.
+    public fun devnet_groth16_vk(): Groth16VerificationKey {
+        Groth16VerificationKey {
+            alpha_g1: x"6d1c152d2705e35fe7a07a66eb8a10a7f42f1e38c412fbbc3ac7f9affc25dc24",
+            beta_g2: x"e20a834c55ae6e2fcbd66636e09322727f317aff8957dd342afa11f936ef7c02cfdc8c9862849a0442bcaa4e03f45343e8bf261ef4ab58cead2efc17100a3b16",
+            gamma_g2: x"edf692d95cbdde46ddda5ef7d422436779445c5e66006a42761e1f12efde0018c212f3aeb785e49712e7a9353349aaf1255dfb31b7bf60723a480d9293938e19",
+            delta_g2: x"98c9283068e4bfc51266dcbabffb56bebeb65ece8d9104609026d0d89187961d0c69a4688b23f8a813ee74349785d116aedfcf3f3de15d7c9123b32eba326f23",
+            gamma_abc_g1: vector[
+                x"29f65be8be6b13c84c1b29d219f35b998db14be4f7506fff4a475512ef0d959f",
+                x"1ddc291dfd35684b634f03cda96ae18139db1653471921c555b2750cbf49908c",
+            ],
+        }
+    }
+
+    public fun default_devnet_configuration(): Configuration {
+        // TODO(zkid): Put reasonable defaults here.
+        Configuration {
+            max_zkid_signatures_per_txn: 3,
+            max_exp_horizon: 100_255_944, // 1159.55 days
+            training_wheels_pubkey: option::some(x"aa"),
+        }
+    }
+
+    // Sets the zkID Groth16 verification key, only callable via governance proposal.
+    // WARNING: If a malicious key is set, this would lead to stolen funds.
+    public entry fun update_groth16_verification_key(fx: &signer, alpha_g1: vector<u8>, beta_g2: vector<u8>, gamma_g2: vector<u8>, delta_g2: vector<u8>, gamma_abc_g1: vector<vector<u8>>) acquires Groth16VerificationKey {
+        system_addresses::assert_aptos_framework(fx);
+
+        if (exists<Groth16VerificationKey>(signer::address_of(fx))) {
+            let Groth16VerificationKey {
+                alpha_g1: _,
+                beta_g2: _,
+                gamma_g2: _,
+                delta_g2: _,
+                gamma_abc_g1: _
+            } = move_from<Groth16VerificationKey>(signer::address_of(fx));
+        };
+
+        let vk = new_groth16_verification_key(alpha_g1, beta_g2, gamma_g2, delta_g2, gamma_abc_g1);
+        move_to(fx, vk);
+    }
+
+    // Sets the zkID configuration, only callable via governance proposal.
+    // WARNING: If a malicious key is set, this would lead to stolen funds.
+    public entry fun update_configuration(fx: &signer, max_zkid_signatures_per_txn: u16, max_exp_horizon: u64, training_wheels_pubkey: Option<vector<u8>>) acquires Configuration {
```

Comment:
probably does not need to be `entry`. can be very awkward to specify a `Option<vector<u8>>` in cmdline.

Reply:
Oh, I guess none of the governance functions need `entry` then!